File: pipeline/src/coast_to_cup/accuracy.py
"""Model-vs-reality stats, deliberately framed as descriptive, not predictive.

The burden model captures only travel, body clock, altitude and heat. It knows
nothing about squad strength, and football has draws, so over a single group
stage (a small, uncontrolled sample) these numbers are exploratory. Published
travel-fatigue effects only separate from noise across thousands of games.
"""

from __future__ import annotations

import math


def _pearson(xs: list[float], ys: list[float]) -> float | None:
    n = len(xs)
    if n < 3:
        return None
    mx, my = sum(xs) / n, sum(ys) / n
    sxy = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    sxx = sum((x - mx) ** 2 for x in xs)
    syy = sum((y - my) ** 2 for y in ys)
    if sxx == 0 or syy == 0:
        return None
    return round(sxy / math.sqrt(sxx * syy), 3)


def summarize(edges: list[dict]) -> dict:
    """Compare the per-match burden edge against actual played results."""
    played = [e for e in edges if e["score_a"] is not None]
    decided = [e for e in played if e["winner"] is not None]
    correct = sum(1 for e in decided if e["edge_team"] == e["winner"])

    # signed: does a bigger burden edge go with a better goal difference?
    margins, goal_diffs = [], []
    for e in played:
        edge_is_a = e["edge_team"] == e["a"]["team_id"]
        gd = (e["score_a"] - e["score_b"]) if edge_is_a else (e["score_b"] - e["score_a"])
        margins.append(e["edge_margin"])
        goal_diffs.append(gd)

    return {
        "n_matches": len(edges),
        "n_played": len(played),
        "n_decided": len(decided),
        "edge_correct": correct,
        "hit_rate": round(correct / len(decided), 3) if decided else None,
        "corr_edge_goaldiff": _pearson(margins, goal_diffs),
        "note": "descriptive only; ignores squad strength; small, uncontrolled sample",
    }

File: web/src/lib/accuracy.ts
// Mirrors the Python accuracy summary so the web can recompute it live.
// Descriptive only: ignores squad strength, small uncontrolled sample.

export interface AccRow {
  edgeTeam: string
  aId: string
  bId: string
  margin: number
  sa: number | null
  sb: number | null
  winner: string | null // pens-aware for knockout games
}

export interface AccSummary {
  nPlayed: number
  nDecided: number
  correct: number
  hitRate: number | null
  corr: number | null
}

function pearson(xs: number[], ys: number[]): number | null {
  const n = xs.length
  if (n < 3) return null
  const mx = xs.reduce((a, b) => a + b, 0) / n
  const my = ys.reduce((a, b) => a + b, 0) / n
  let sxy = 0
  let sxx = 0
  let syy = 0
  for (let i = 0; i < n; i++) {
    const dx = xs[i] - mx
    const dy = ys[i] - my
    sxy += dx * dy
    sxx += dx * dx
    syy += dy * dy
  }
  if (sxx === 0 || syy === 0) return null
  return sxy / Math.sqrt(sxx * syy)
}

export function summarize(rows: AccRow[]): AccSummary {
  const played = rows.filter((r) => r.sa != null && r.sb != null)
  const decided = played.filter((r) => r.winner !== null)
  const correct = decided.filter((r) => r.winner === r.edgeTeam).length

  const margins: number[] = []
  const goalDiffs: number[] = []
  for (const r of played) {
    const gd = r.edgeTeam === r.aId ? r.sa! - r.sb! : r.sb! - r.sa!
    margins.push(r.margin)
    goalDiffs.push(gd)
  }

  return {
    nPlayed: played.length,
    nDecided: decided.length,
    correct,
    hitRate: decided.length ? correct / decided.length : null,
    corr: pearson(margins, goalDiffs),
  }
}
